### PR TITLE
KEYCLOAK-19726 Fix return types of ModelCriteriaBuilder methods

### PR DESCRIPTION
**Pull request by hmlnarik**
Affects #8571 and #8615

### CODE REVIEW

**Review by vramik — DISMISSED** [review]
LGTM thanks @hmlnarik

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @hmlnarik. I added few comments, could you please have a look.

File: model/map/src/main/java/org/keycloak/models/map/loginFailure/MapUserLoginFailureProvider.java [hunk 1]
```diff
@@ -54,9 +54,9 @@ public MapUserLoginFailureProvider(KeycloakSession session, MapStorage<MapUserLo
 
     @Override
     public UserLoginFailureModel getUserLoginFailure(RealmModel realm, String userId) {
-        ModelCriteriaBuilder<UserLoginFailureModel> mcb = criteria();
-        mcb = mcb.compare(UserLoginFailureModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
-                .compare(UserLoginFailureModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, userId);
+        DefaultModelCriteria<UserLoginFailureModel> mcb = criteria();
+        mcb = mcb.compare(UserLoginFailureModel.SearchableFields.REALM_ID, DefaultModelCriteria.Operator.EQ, realm.getId())
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/loginFailure/MapUserLoginFailureProvider.java:58]: Just for this reason we should keep `ModelCriteriaBuilder.Operator`

File: model/map/src/main/java/org/keycloak/models/map/storage/criteria/DefaultModelCriteria.java [hunk 2]
```diff
@@ -63,18 +63,18 @@ private DefaultModelCriteria(ModelCriteriaNode<M> node) {
     }
 
     @Override
-    public DefaultModelCriteria<M> and(ModelCriteriaBuilder<M>... mcbs) {
-        if (mcbs.length == 1) {
-            ModelCriteriaNode<M> toBeChild = ((DefaultModelCriteria<M>) mcbs[0]).node;
+    public DefaultModelCriteria<M> and(DefaultModelCriteria<M>... mcbs) {
+        if (mcbs.length == 1 && isEmpty()) {
+            ModelCriteriaNode<M> toBeChild = mcbs[0].node;
             if (toBeChild.getNodeOperator() == ExtOperator.AND || toBeChild.getNodeOperator() == ExtOperator.OR) {
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/criteria/DefaultModelCriteria.java:69]: Good question, and the truth is we should do it always. Thanks!

File: model/map/src/main/java/org/keycloak/models/map/storage/criteria/ModelCriteriaNode.java [hunk 3]
```diff
@@ -35,45 +36,45 @@
...
                     return null;
                 }
-                final ModelCriteriaBuilder[] operands = node.getChildren().stream()
+                final C[] operands = node.getChildren().stream()
                   .map(n -> n.flashToModelCriteriaBuilder(mcb))
                   .filter(Objects::nonNull)
-                  .toArray(ModelCriteriaBuilder[]::new);
-                return operands.length == 0 ? null : (C) mcb.and(operands);
+                  .toArray(n -> (C[]) Array.newInstance(mcb.getClass(), n));
+                return operands.length == 0 ? null : mcb.and(operands);
```
Comment (vramik) [inline model/map/src/main/java/org/keycloak/models/map/storage/criteria/ModelCriteriaNode.java:47]: During rebase I've debugged the code and at this point operands contains two identical instances of the same MCB, I was wondering if it is expected. It seems like bug to me, but maybe there is something wrong in my end. It ends up calling ModelCriteriaBuilder.and method where there are two identical builders. I'll perform some additional tests to investigate it further. In the meantime I wanted to let you know about it. Thank you

Comment (mhajas) [reply model/map/src/main/java/org/keycloak/models/map/storage/criteria/ModelCriteriaNode.java:47]: @vramik, is this happening also after the latest changes from @hmlnarik? I can imagine it was caused by not copying the tree in `and` and `or` methods, but it is not there anymore.

Comment (vramik) [reply model/map/src/main/java/org/keycloak/models/map/storage/criteria/ModelCriteriaNode.java:47]: @mhajas it is happening with JpaClientModelCriteriaBuilder from jpa store work and thanks to @hmlnarik it turned out it is caused by the fact I was returning `this` there. We've discussed it and now I'm going to rewrite my code. So this comment chould be resolved.

**Review by mhajas — APPROVED** [review]
Thanks for the fixes @hmlnarik. Approving.

**Review by vramik — APPROVED** [review]

**Pull request merged**